Plugin/Catalog/Block/AbstractBlock.php

```diff
@@ -0,0 +1,260 @@
+    public $nostoHelperAccount;
+
+    /** @var NostoLogger */
+    public $logger;
+
+    /**
+     * AbstractBlock constructor.
+     * @param Context $context
+     * @param Http $httpRequest
+     * @param NostoCmpHelperData $nostoCmpHelperData
+     * @param NostoHelperAccount $nostoHelperAccount
+     * @param NostoLogger $logger
+     */
+    public function __construct(
+        Context $context,
+        Http $httpRequest,
+        NostoCmpHelperData $nostoCmpHelperData,
+        NostoHelperAccount $nostoHelperAccount,
+        NostoLogger $logger
+    ) {
+        $this->httpRequest = $httpRequest;
+        $this->nostoCmpHelperData = $nostoCmpHelperData;
+        $this->nostoHelperAccount = $nostoHelperAccount;
+        $this->logger = $logger;
+        $this->storeManager = $context->getStoreManager();
+        parent::__construct($context);
+    }
+
+    /**
+     * Checks if current sorting order is Nosto's `Personalized for you`
+     * and category sorting is enabled
+     *
+     * @return bool
+     */
+    public function isCmpCurrentSortOrder()
+    {
+        try {
+            /* @var Store $store */
+            $store = $this->storeManager->getStore();
+        } catch (NoSuchEntityException $e) {
+            $this->logger->info('Cannot get store');
+            return false;
+        }
+
+        $currentOrder = $this->getCurrentOrder();
+        if ($currentOrder !== null
+            && $currentOrder === NostoHelperSorting::NOSTO_PERSONALIZED_KEY
+            && $this->nostoHelperAccount->nostoInstalledAndEnabled($store)
+            && $this->nostoCmpHelperData->isCategorySortingEnabled($store)
+        ) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * In case CMP is selected as sorting order
+     * and result is not empty
+     *
+     * @return bool
+     */
+    public function isCmpTakingOverCatalog()
+    {
+        if ($this->isCmpCurrentSortOrder() && self::$totalProducts !== null) {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @return string|null
+     */
+    private function getCurrentOrder()
+    {
+        return $this->httpRequest->getParam('product_list_order');
+    }
+
+    /**
+     * @param int $totalProducts
+     */
+    public function setTotalProducts($totalProducts)
+    {
+        self::$totalProducts = $totalProducts;
+    }
+
+    /**
+     * @return int
+     */
+    public function getTotalProducts()
+    {
+        return self::$totalProducts;
+    }
+
+    /**
+     * @param MagentoToolbar|MagentoPager $block
+     * @param $result
+     * @return float|int
+     */
+    public function afterGetFirstNum($block, $result)
+    {
+        if ($this->isCmpTakingOverCatalog()) {
+            $pageSize = $block->getCollection()->getPageSize();
+            $currentPage = $this->getCurrentPageNumber();
+            return $pageSize * ($currentPage - 1) + 1;
+        }
+        return $result;
+    }
```

Comment:
I don't understand this method. Does it return the number of products on a page?